### ISSUE TITLE
HTTP API Add Crashes Daemon

### ISSUE THREAD

**Issue by nycoliver** [ipfs/go-ipfs#3336]
#### Version information:

```
go-ipfs version: 0.4.5-dev-17c629d
Repo version: 4
System version: amd64/darwin
Golang version: go1.7.3
```

#### Type:

Bug

#### Priority:

P3
#### Description:

HTTP API requests almost always fail with the new go-ipfs. They occasionally work, but most of the time daemon throws the following errors. After enough (~100) requests, the daemon consistently crashes.

```
17:07:01.242 ERROR commands/h: err: multipart: NextPart: EOF handler.go:288
17:07:01.474 ERROR commands/h: err: multipart: NextPart: http: invalid Read on closed Body handler.go:288
17:07:01.626 ERROR commands/h: err: multipart: Part Read: http: invalid Read on closed Body handler.go:288
17:07:01.803 ERROR commands/h: err: multipart: NextPart: http: invalid Read on closed Body handler.go:288
17:07:02.212 ERROR commands/h: err: multipart: NextPart: http: invalid Read on closed Body handler.go:288
panic: runtime error: index out of range

goroutine 2946 [running]:
panic(0x7e1ba0, 0xc4200100f0)
    /usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
github.com/ipfs/go-ipfs/core/coreunix.(*Adder).Finalize(0xc42292d200, 0xc420078020, 0xcfb3c0, 0xc420078020, 0xc420078001)
    /Users/o/Code/gocode/src/github.com/ipfs/go-ipfs/core/coreunix/add.go:198 +0x2cb
github.com/ipfs/go-ipfs/core/commands.glob..func6.1(0xd0a3a0, 0xc4236b7500, 0xc4236b7500, 0x41bf8)
    /Users/o/Code/gocode/src/github.com/ipfs/go-ipfs/core/commands/add.go:211 +0x136
github.com/ipfs/go-ipfs/core/commands.glob..func6.2(0xc423673c20, 0xc4240f5380, 0xd0e340, 0xc423c6c780, 0xd0d700, 0xc42361c930)
    /Users/o/Code/gocode/src/github.com/ipfs/go-ipfs/core/commands/add.go:225 +0x7c
created by github.com/ipfs/go-ipfs/core/commands.glob..func6
    /Users/o/Code/gocode/src/github.com/ipfs/go-ipfs/core/commands/add.go:230 +0x7fb
```

Reproduce with the following test. Make sure to set CORS. Press add button repeatedly and the daemon will eventually crash.

``` html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <script src="https://unpkg.com/ipfs-api@11.0.1/dist/index.js"></script>
  </head>
  <body>
    <button onclick='addToIPFS()'>Add</button>
  </body>
  <script>
  var ipfs = window.IpfsApi();

  function addToIPFS() {
    ipfs.util.addFromURL('https://avatars1.githubusercontent.com/u/2553061?v=3&s=466', function(err, res) {
        if (err || res === null || res.length === 0) {
          console.error("Failed to add file", err);
          return;
        }

        console.log(res);
    });
  }
  </script>
</html>
```

I think this is caused by https://github.com/ipfs/go-ipfs/commit/68d8a298c58ed950bd44118cc5e6d1ffc2def395 -- issue only appears after that commit.

**Comment by whyrusleeping** [comment]
@nycoliver Thanks for pointing this out... Very odd that this doesnt occur while using the command line `ipfs add`.

Any other insights into what the root of the issue is? 68d8a29 seemed pretty inoccuous to me. I'll have to go do some more thinking about it.

**Comment by whyrusleeping** [comment]
@nycoliver it appears the panic was caused by the removal of flush from the end of the function. The errors go away if you remove the flush from the top of the for loop. Youre right about that commit being the issue, but i honestly have no idea why at this point.

**Comment by whyrusleeping** [comment]
Current train of thought: It might be something weird to do with the http lib Flush call?

https://golang.org/src/net/http/server.go#L1399

**Comment by whyrusleeping** [comment]
cc @keks any idea here?

**Comment by whyrusleeping** [comment]
@nycoliver @diasdavid One more question, does this only happen with the js-ipfs-api? Or can someone reproduce the issue with curl?

**Comment by keks** [comment]
So I guess `Flush`ing before all content is there triggers bugs where readers don't check if all data has been read. It not only flushes the HTTP connection but also the local reader. This also makes sense because apparently the [`rootNode`]'s links have not been populated - at least that's how I read the stack trace.

I'll try to dig deeper.

[`rootNode`]:https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L198

**Comment by keks** [comment]
Hmm, I can't really get this to reproduce so it's hard for me to debug. I have two ideas right now: Either [`Parse()`](https://github.com/ipfs/go-ipfs/blob/master/commands/http/parse.go#L16) has a hickup when the request body reader flushes before being done, or the JS API client stops sending stuff when it received a header.
Does HTTP allow sending after receiving a request in non-ws mode? If not, the error is definitely on my side and I need to reconsider the flushing, maybe make it conditional.

**Comment by keks** [comment]
Okay so I could reproduce now. The add request seems to be sent complely, so there doesn't seem to be a problem on the js-ipfs-api side.
![wireshark](https://cloud.githubusercontent.com/assets/146678/20013885/22dcda5e-a2b5-11e6-8ac8-f102724711ce.png)

**Comment by keks** [comment]
```
var ErrBodyReadAfterClose = errors.New("http: invalid Read on closed Body")
```

ErrBodyReadAfterClose is returned when reading a Request or Response Body after the body has been closed. This typically happens when the body is read after an HTTP Handler calls WriteHeader or Write on its ResponseWriter.

._.

**Comment by keks** [comment]
That commit at least fixes this problem, but I suspect there are more issues of this kind in this part. I'll take another look next week

**Comment by whyrusleeping** [comment]
@nycoliver mind testing with @keks commit?

**Comment by whyrusleeping** [comment]
@keks are you able to reproduce the issue with curl? I'd love to turn this into a sharness test somehow.

If you cant do it with curl, then writing a small go program works, we can put it in our test dependencies folder.

**Comment by keks** [comment]
hmm, I tried `curl -v -F foo=@/tmp/phr/1.txt http://localhost:5001/api/v0/add` but gives a different error (`ERROR commands/h: err: multipart: NextPart: unexpected EOF handler.go:288`) and sometimes works. Also it doesn't crash the daemon. Looking at the wireshark output it seems like all three HTTP requests are very similar (different client headers and that kind of stuff), so I assume it's a matter of timing. Trying to figure out what timing yields which result is tedious and I am not sure it's worthwhile - instead I would propose finding a way to make the result the same for all kinds of timing.

One idea I had for that is making `w.Flush`, `w.Write` and `w.WriteHeader` block until `r.Body` `EOF`'d. I am afraid that this might result in deadlocks somewhere but I don't know the codebase enough. What do you think about that?

**Comment by hsanjuan** [comment]
@keks try `curl -0` instead. I reported this at #3332

**Comment by keks** [comment]
@hsanjuan Thanks for pointing that out. Interestingly there is no error now using curl. I guess it's just a matter of timing or how large the initial TCP segments are. Anyway, the ipfs daemon shouldn't care about that.
I implemented the idea I was writing about in my last comment but unfortunately now API requests block indefinitely.

I just realized I only mentioned this on IRC: There is a more general problem here, take a look at #3367. I guess I need to do more restructuring here.
